File: runners/python/steps/retry_steps.py
from behave import given, when, then
from assertpy import assert_that
import math


# Retry assertion steps
@then('the request should have been retried {count:d} times')
def step_request_retried_count(context, count):
    actual_retries = context.test_context.get_retry_count()
    assert_that(actual_retries).is_equal_to(count)


@then('the request should not have been retried')
def step_request_not_retried(context):
    retry_count = context.test_context.get_retry_count()
    assert_that(retry_count).is_equal_to(0)


@then('the final attempt should succeed')
def step_final_attempt_succeed(context):
    context.test_context.assert_response_success()
    assert_that(context.test_context.last_response).is_not_none()


@then('the retry delays should follow exponential backoff')
def step_retry_delays_exponential(context):
    delays = context.test_context.get_retry_delays()
    assert_that(delays).is_not_empty()
    
    # Verify exponential growth pattern
    for i in range(1, len(delays)):
        assert_that(delays[i]).is_greater_than(delays[i - 1])


@then('the retry delays should be linear with base delay {base_delay:d}ms')
def step_retry_delays_linear(context, base_delay):
    delays = context.test_context.get_retry_delays()
    assert_that(delays).is_not_empty()
    
    # Verify linear growth pattern
    for i, delay in enumerate(delays):
        expected_delay = base_delay * (i + 1)
        tolerance = base_delay * 0.1  # 10% tolerance
        assert_that(delay).is_close_to(expected_delay, tolerance)


@then('all retry attempts should have failed')
def step_all_retry_attempts_failed(context):
    context.test_context.assert_response_failure()
    retry_count = context.test_context.get_retry_count()
    assert_that(retry_count).is_greater_than(0)


@then('the circuit breaker should be triggered after {threshold:d} failures')
def step_circuit_breaker_triggered(context, threshold):
    assert_that(context.test_context.circuit_breaker_triggered).is_true()
    assert_that(context.test_context.circuit_breaker_failure_count).is_equal_to(threshold)


@then('subsequent requests should fail fast without retries')
def step_subsequent_requests_fail_fast(context):
    assert_that(context.test_context.circuit_breaker_open).is_true()
    retry_count = context.test_context.get_retry_count()
    assert_that(retry_count).is_equal_to(0)


@then('the configuration should fail with validation error')
def step_configuration_validation_error(context):
    assert_that(context.test_context.last_error).is_not_none()
    error_message = str(context.test_context.last_error)
    assert_that(error_message).contains('validation')


@then('the retry delays should include random jitter')
def step_retry_delays_include_jitter(context):
    delays = context.test_context.get_retry_delays()
    assert_that(len(delays)).is_greater_than(1)
    
    # Verify that delays are not exactly exponential (indicating jitter)
    has_jitter = False
    for i in range(1, len(delays)):
        expected_exponential = delays[0] * (2 ** i)
        tolerance = expected_exponential * 0.3  # 30% tolerance for jitter
        if abs(delays[i] - expected_exponential) > tolerance * 0.1:
            has_jitter = True
            break
    
    assert_that(has_jitter).is_true()

File: runners/java/src/test/java/dev/openfga/sdk/conformance/steps/RetrySteps.java
package dev.openfga.sdk.conformance.steps;

import io.cucumber.java.en.Then;
import org.assertj.core.api.Assertions;
import dev.openfga.sdk.conformance.support.TestContext;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;

public class RetrySteps {
    
    private final TestContext testContext;

    public RetrySteps(TestContext testContext) {
        this.testContext = testContext;
    }

    @Then("the request should have been retried {int} times")
    public void theRequestShouldHaveBeenRetriedTimes(int expectedRetries) {
        int actualRetries = getRetryCount();
        Assertions.assertThat(actualRetries).isEqualTo(expectedRetries);
    }

    @Then("the request should not have been retried")
    public void theRequestShouldNotHaveBeenRetried() {
        int retryCount = getRetryCount();
        Assertions.assertThat(retryCount).isEqualTo(0);
    }

    @Then("the final attempt should succeed")
    public void theFinalAttemptShouldSucceed() {
        Assertions.assertThat(testContext.getLastError()).isNull();
        Assertions.assertThat(testContext.getLastResponse()).isNotNull();
    }

    @Then("the retry delays should follow exponential backoff")
    public void theRetryDelaysShouldFollowExponentialBackoff() {
        List<Integer> delays = getRetryDelays();
        Assertions.assertThat(delays).isNotEmpty();
        
        // Verify exponential growth pattern
        for (int i = 1; i < delays.size(); i++) {
            Assertions.assertThat(delays.get(i)).isGreaterThan(delays.get(i - 1));
        }
    }

    @Then("the retry delays should be linear with base delay {int}ms")
    public void theRetryDelaysShouldBeLinearWithBaseDelay(int baseDelay) {
        List<Integer> delays = getRetryDelays();
        Assertions.assertThat(delays).isNotEmpty();
        
        // Verify linear growth pattern
        for (int i = 0; i < delays.size(); i++) {
            int expectedDelay = baseDelay * (i + 1);
            double tolerance = baseDelay * 0.1; // 10% tolerance
            Assertions.assertThat(delays.get(i)).isBetween(
                (int)(expectedDelay - tolerance), 
                (int)(expectedDelay + tolerance)
            );
        }
    }

    @Then("all retry attempts should have failed")
    public void allRetryAttemptsShouldHaveFailed() {
        Assertions.assertThat(testContext.getLastError()).isNotNull();
        int retryCount = getRetryCount();
        Assertions.assertThat(retryCount).isGreaterThan(0);
    }

    @Then("the circuit breaker should be triggered after {int} failures")
    public void theCircuitBreakerShouldBeTriggeredAfterFailures(int threshold) {
        boolean circuitBreakerTriggered = testContext.getSavedData().containsKey("circuit_breaker_triggered") &&
                                         (Boolean) testContext.getSavedData().get("circuit_breaker_triggered");
        Assertions.assertThat(circuitBreakerTriggered).isTrue();
        
        int failureCount = testContext.getSavedData().containsKey("circuit_breaker_failure_count") ?
                          (Integer) testContext.getSavedData().get("circuit_breaker_failure_count") : 0;
        Assertions.assertThat(failureCount).isEqualTo(threshold);
    }

    @Then("subsequent requests should fail fast without retries")
    public void subsequentRequestsShouldFailFastWithoutRetries() {
        boolean circuitBreakerOpen = testContext.getSavedData().containsKey("circuit_breaker_open") &&
                                   (Boolean) testContext.getSavedData().get("circuit_breaker_open");
        Assertions.assertThat(circuitBreakerOpen).isTrue();
        
        int retryCount = getRetryCount();
        Assertions.assertThat(retryCount).isEqualTo(0);
    }

    @Then("the configuration should fail with validation error")
    public void theConfigurationShouldFailWithValidationError() {
        Assertions.assertThat(testContext.getLastError()).isNotNull();
        Assertions.assertThat(testContext.getLastError().getMessage()).contains("validation");
    }

    @Then("the retry delays should include random jitter")
    public void theRetryDelaysShouldIncludeRandomJitter() {
        List<Integer> delays = getRetryDelays();
        Assertions.assertThat(delays.size()).isGreaterThan(1);
        
        // Verify that delays are not exactly exponential (indicating jitter)
        boolean hasJitter = false;
        for (int i = 1; i < delays.size(); i++) {
            double expectedExponential = delays.get(0) * Math.pow(2, i);
            double tolerance = expectedExponential * 0.3; // 30% tolerance for jitter
            if (Math.abs(delays.get(i) - expectedExponential) > tolerance * 0.1) {
                hasJitter = true;
                break;
            }
        }
        Assertions.assertThat(hasJitter).isTrue();
    }

    @SuppressWarnings("unchecked")
    private int getRetryCount() {
        if (testContext.getSavedData().containsKey("retry_attempts")) {
            List<Long> attempts = (List<Long>) testContext.getSavedData().get("retry_attempts");
            return attempts.size();
        }
        return 0;
    }

    @SuppressWarnings("unchecked")
    private List<Integer> getRetryDelays() {
        if (testContext.getSavedData().containsKey("retry_delays")) {
            return (List<Integer>) testContext.getSavedData().get("retry_delays");
        }
        return new ArrayList<>();
    }
}
